### PR TITLE
feat: strict eslint

### PR DESCRIPTION
**Pull request by Benehiko**
## Related Issue or Design Document

This PR adds stricter type checking to the code base. This allows us to catch bugs early in the library as well as in playwright tests.

Tests were also added to some of the components to ensure they behave as expected, such as the error messages. Translation tests now also checks that template strings exist within the translation based on the english version.

## Checklist

- [ ] I have read the [contributing guidelines](../blob/master/CONTRIBUTING.md) and signed the CLA.
- [ ] I have referenced an issue containing the design document if my change introduces a new feature.
- [ ] I have read the [security policy](../security/policy).
- [ ] I confirm that this pull request does not address a security vulnerability. 
      If this pull request addresses a security vulnerability, 
      I confirm that I got approval (please contact [security@ory.sh](mailto:security@ory.sh)) from the maintainers to push the changes.
- [ ] I have added tests that prove my fix is effective or that my feature works.
- [ ] I have added the necessary documentation within the code base (if appropriate).

### CODE REVIEW

File: src/markup-components/component-wrapper.tsx [hunk 1]
```diff
@@ -1,18 +1,18 @@
...
 
-export type Context = {
+export interface Context {
   locale?: keyof typeof locales
 }
 
-export const ComponentWrapper = <Props extends {}>(
-  Component: FunctionComponent<Props>,
-  props: Props,
+export const ComponentWrapper = <WrapperProps extends PropsWithChildren>(
```
Comment (Benehiko) [inline src/markup-components/component-wrapper.tsx:13]: eslint strict types don't like the usage of a type extending an empty object (any).

Not sure if this is the correct solution to this. Another option might be to exclude this file from eslint?

File: src/markup-components/express.spec.ts
Comment (Benehiko) [inline src/markup-components/express.spec.ts:1]: this file isn't being run yet since playwright complains about vanilla-extract stuff since it's not correctly bundling the code for it for some reason... still need to figure out why tho.

File: src/markup-components/express.ts [hunk 2]
```diff
@@ -34,7 +34,11 @@ export const RegisterOryElementsExpress: OryEelementsExpressRoute = (
       // merge also creates a new object ensuring that we don't pollute the default or request theme
       inlineTheme = merge({}, inlineTheme, theme)
     }
-    res.send(`body {${assignInlineVars(oryTheme, inlineTheme).toString()}}`)
+    res.send(
+      `body {${Object.entries(
+        assignInlineVars(oryTheme, inlineTheme),
+      ).toString()}}`,
+    )
```
Comment (Benehiko) [inline src/markup-components/express.ts:41]: TODO: still need to test if this doesn't break something

File: src/react-components/ory/helpers/error-messages.tsx [hunk 3]
```diff
@@ -59,12 +59,12 @@ export const NodeMessages = ({
     nodeMessage({ message, key: `ui-messsage-${message.id}-${key}` }),
   )
 
-  const $allMessages = [...($groupMessages || []), ...($messages || [])]
+  const $allMessages = [...($groupMessages ?? []), ...($messages ?? [])]
```
Comment (Benehiko) [inline src/react-components/ory/helpers/error-messages.tsx:62]: error messages component needs some tests!

File: src/theme/message.css.ts [hunk 4]
```diff
@@ -14,12 +14,12 @@ const SeverityTypes = [
...
 export type Severity = (typeof SeverityTypes)[number]
 
-type MessageVariants = {
-  textPosition: Record<string, any>
-  severity: Record<Severity, any>
-}
+// interface MessageVariants {
+//   textPosition: Record<string, unknown>
+//   severity: Record<Severity, unknown>
+// }
```
Comment (Benehiko) [inline src/theme/message.css.ts:20]: todo remove? 
I don't think it's necessary to have this type here.

File: src/markup/express.ts
Comment (Benehiko) [inline src/markup/express.ts:1]: this file wasn't used? it's a duplicate of what we have inside the src/markup-components.

File: .eslintignore [hunk 5]
```diff
@@ -1,2 +1,5 @@
 node_modules
-dist
\ No newline at end of file
+dist
+src/markup-components/component-wrapper.tsx
+src/markup-components/components.ts
+src/markup-components/express.ts
```
[the remark this answers is not in the document]
Comment (jonas-jonas) [reply .eslintignore:5]: But why can't the files be linted?

File: src/react-components/button-link.tsx [hunk 6]
```diff
@@ -9,7 +9,7 @@ import {
   buttonLinkStyle,
 } from "../theme/button-link.css"
 
-export type CustomHref = {
+export interface CustomHref {
```
Comment (jonas-jonas) [inline src/react-components/button-link.tsx:12]: Why this change?

File: src/react-components/ory/helpers/node.tsx
Comment (jonas-jonas) [inline src/react-components/ory/helpers/node.tsx:1]: Lots of seemingly unchecked type casts in this file. Would it be possible to type check the values before casting?

File: src/react-components/tests/translations.spec.ts [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+  },
+})
+
+test("language keys and templates match", async ({
+  en,
+  translations,
+  templates,
+}) => {
+  await test.step("Checking language keys", () => {
+    for (const [, translation] of Object.entries(translations)) {
```
Comment (jonas-jonas) [inline src/react-components/tests/translations.spec.ts:53]:
```suggestion
    for (const translation of Object.values(translations)) {
```

**Review by jonas-jonas — COMMENTED** [review]
Nice, just a few comments.

File: .eslintignore (hunk 5 above)
[the remark this answers is not in the document]
Comment (Benehiko) [reply .eslintignore:5]: they can, but they require some changes. I don't want to touch them right now since I don't have tests in elements to verify that the changes worked.

File: src/react-components/button-link.tsx (hunk 6 above)
Comment (Benehiko) [reply src/react-components/button-link.tsx:12]: eslint prefers interfaces over type definitions. since it changed the others since they were being used by other interfaces, this type was changed to reflect a consistent code-base. eslint is quite opinionated and it's not really necessarily about a technical reason, more about formatting in most cases.


https://typescript-eslint.io/rules/consistent-type-definitions/

Comment (jonas-jonas) [reply src/react-components/button-link.tsx:12]: I'd prefer types then :D But consistency is good. Thanks for the explanation.

**Review by mszekiel — APPROVED** [review]
Looks good, take a look on the assertions again.

**Review by jonas-jonas — APPROVED** [review]

File: src/react-components/ory/helpers/node.tsx [hunk 8]
```diff
@@ -56,47 +56,131 @@ export const getNodeLabel = (node: UiNode): UiText | undefined => {
...
+      }
+      return {
+        ...accumulator,
+        [key]: value as string | number,
+      }
+    },
+    {},
+  )
+
+  return intl.formatMessage(
```
Comment (Benehiko) [inline src/react-components/ory/helpers/node.tsx:176]: This needs some review. I am changing the date format as well here to only have 2 decimal places in the minutes we display. For example "the flow has expired 10.00 minutes ago".

Right now it's a bit unwieldy and would display "the flow has expired -10.0004123 minutes ago".

Comment (Benehiko) [reply src/react-components/ory/helpers/node.tsx:176]: There are some tests inside the error-messages.spec file and inside the node.spec.tsx file.

Comment (aeneasr) [reply src/react-components/ory/helpers/node.tsx:176]: Probably makes more sense to fix this in kratos?

Comment (Benehiko) [reply src/react-components/ory/helpers/node.tsx:176]: kratos gives back a unix timestamp from where it expired. So the date. 
The ui then has to render the time or the minutes from/until that date.

Comment (aeneasr) [reply src/react-components/ory/helpers/node.tsx:176]: Ah I get it, this basically formats the timestamp in separate keys (until, since, ...). Makes sense to me

Comment (Benehiko) [reply src/react-components/ory/helpers/node.tsx:176]: yeah exactly. Do you think it's fine to merge?

**Review by aeneasr — APPROVED** [review]
